src/download/mod.rs

```diff
@@ -0,0 +1,342 @@
+pub mod package;
+
+use std::borrow::Cow;
+use std::io::{BufReader, Read};
+use std::ffi::OsStr;
+use std::fs::File;
+use std::fs;
+use std::path::Path;
+use std::str::FromStr;
+use std::time::Duration;
+
+use anyhow::{Context, Result, anyhow, bail};
+use globset::GlobSet;
+use hard_xml::XmlRead;
+use log::{debug, info, warn};
+use omaha::FileSize;
+use reqwest::{StatusCode, blocking::Client, redirect::Policy};
+use sha2::digest::DynDigest;
+use url::Url;
+
+use crate::{Package, PackageStatus};
+
+const DOWNLOAD_TIMEOUT: u64 = 3600;
+const HTTP_CONN_TIMEOUT: u64 = 20;
+const MAX_DOWNLOAD_RETRY: u32 = 20;
+
+pub const TARGET_FILENAME_DEFAULT: &str = "oem-azure.gz";
+pub const PAYLOAD_URL_DEFAULT: &str = "https://update.release.flatcar-linux.net/amd64-usr/current/oem-azure.gz";
+
+const UNVERFIED_SUFFIX: &str = ".unverified";
+const TMP_SUFFIX: &str = ".tmp";
+
+pub struct DownloadResult {
+    pub hash_sha256: omaha::Hash<omaha::Sha256>,
+    pub hash_sha1: omaha::Hash<omaha::Sha1>,
+    pub data: File,
+}
+
+pub fn hash_on_disk<T: omaha::HashAlgo>(path: &Path, maxlen: Option<usize>) -> Result<omaha::Hash<T>> {
+    let file = File::open(path).context(format!("File::open({:?})", path))?;
+    let mut hasher = T::hasher();
+
+    let filelen = file.metadata().context(format!("failed to get metadata of {:?}", path))?.len() as usize;
+
+    let mut maxlen_to_read: usize = match maxlen {
+        Some(len) => {
+            if filelen < len {
+                filelen
+            } else {
+                len
+            }
+        }
+        None => filelen,
+    };
+
+    const CHUNKLEN: usize = 10485760; // 10M
+
+    let mut freader = BufReader::new(file);
+    let mut chunklen: usize;
+    let mut databuf = vec![0u8; 0];
+
+    while maxlen_to_read > 0 {
+        if maxlen_to_read < CHUNKLEN {
+            chunklen = maxlen_to_read;
+        } else {
+            chunklen = CHUNKLEN;
+        }
+
+        databuf.resize(databuf.len() + chunklen, 0);
+
+        freader.read_exact(&mut databuf).context(format!("failed to read_exact(chunklen {chunklen:?})"))?;
+
+        maxlen_to_read -= chunklen;
+
+        hasher.update(&databuf);
+    }
```

Comment:
You could allocate CHUNKLEN up-front and do the truncation only on the last iteration where maxlen_to_read is less than CHUNKLEN.

Also resize inside the loop looks wrong as it grows the databuf with each iteration by chunklen, so on each iteration multiple chunklens will be read from freader into databuf, but maxlen_to_read will be reduced only by one chunklen.

Also, if you care about not allocating CHUNKLEN up-front when maxlen_to_read is zero, you could probably put the loop together with freader and databuf inside `if maxlen_to_read > 0`.

```suggestion
    let mut freader = BufReader::new(file);
    let mut databuf = vec![0u8; CHUNKLEN];

    while maxlen_to_read > 0 {
        if maxlen_to_read < CHUNKLEN {
            // last and submaximal chunk to read, shrink the buffer for it
            databuf.truncate(maxlen_to_read)
        }

        // I have no idea if the format string is correct here.
        freader.read_exact(&mut databuf).context(format!("failed to read_exact(chunklen {databuf.len():?})"))?;

        maxlen_to_read -= databuf.len();

        hasher.update(&databuf);
    }
```

Reply:
Right. Fixed.

Though the format string should be different.
We cannot write a function call `databuf.len()` inline inside `format!`.
The only way is like:
```
format!("failed to read_exact(chunklen {:?})", databuf.len())
```